### ISSUE TITLE
Amcrest (new feature) - Add NVR channel selection support

### ISSUE THREAD

**Issue by GaryOkie** [pnbruckner/homeassistant-config#121]
A hopefully straight-forward enhancement that would allow the component to connect directly to an Amcrest (or possibly Dahua) NVR is to enable a new  option in the HA Amcrest config to select a channel # that is associated with each camera.  (maybe this can already be done via ffmpeg_arguments, but that is an obtuse way to accomplish channel selection, if so).

From the python-amcrest documentation, the API shows an existing channel parameter, (apparently hardcoded to 1?)...
> **realtime_stream**(channel=1, typeno=0, path_file=None)
> 
> **rtsp_url**(channelno=None, typeno=None)
>  Return RTSP streaming url
> 
> Params:
> **channelno**: integer, the video channel index which starts from 1, default 1 if not specified.
> **typeno**: the stream type, default 0 if not specified. It can be the following value:
> 
>            0-Main Stream 1-Extra Stream 1 (Sub Stream) 2-Extra Stream 2 (Sub Stream)

Seeing this doc also brings up an additional feature request unrelated to the NVR support - being able to select the stream (typeno).   Currently the Amcrest component only lets you choose high or low resolution, but being able to select the substream as an alternative for a lower resolution/lower frame rate settings that are defined in the camera for HA streaming offers more flexibility.

Note: This NVR enhancement request had been previously made to @tchellomello a few years ago - [here.](https://github.com/tchellomello/python-amcrest/issues/89).

Let me know if this feature request would be better placed in the main HA github issues.

**Comment by pnbruckner** [comment]
Hi. From what I've seen so far in the amcrest package, some methods allow specifying the channel number, whereas others do not (they just assume the first channel.) Also, I'm not sure what it means to the overall integration (amcrest package and HA component) to handle multiple channels -- do they look like multiple cameras, with full control, just with a non-default channel number? or will only some things work on each channel but not everything? Unfortunately, I don't have an Amcrest NVR, so I can't really experiment to see how it works. (I'm already at a disadvantage having only a couple of Amcrest models, so I'm not 100% sure the things I'm already doing aren't going to break things.)

So, bottom line, I can look into this, but there are a few other, basic things, I'm trying to get done first, at least as far as what I'm officially submitting for the standard amcrest component. Once I get those things squared away, we can try some things with my custom component first.

**Comment by pnbruckner** [comment]
BTW, you can already select the substream. I.e., "high" corresponds to typeno 0 and "low" corresponds to typeno 1.

**Comment by GaryOkie** [comment]
Thanks for reviewing this request, and for clarifying the resolution option already selects substream.     The docs do need to be revised though since they state a specific resolution for high and low which do not necessarily correspond to the substream settings, especially for newer 3MP/4MP/+ cameras.

And full disclosure - I don't have an Amcrest NVR either!    I do have a Dahua NVR , and in every respect the HTTP/CGI calls I've made directly to the cameras work perfectly when directed to the NVR with their respective channel number instead.     Updating the calls to use channel=n, where n defaults to 1 when not specified, would be very unlikely to cause any problems.

My camera POE security network is segregated from my home LAN.  This was the main reason for wanting to connect directly to the NVR rather than the cameras directly.    A minor secondary reason is that authentication and IP assignment is singular to the NVR.

Since I can now connect directly to the cameras instead of the NVR,  this a low priority, but clearly a feature others have asked for.   I know you are working on other things of higher importance.    

The really big deal for me is that I can use existing Amcrest code to drive not just my remote Amcrest IPC's,   but now all the Dahua IPC's at home with all the features (except NVR) I've been needing.   All your great work in improving this component has just been phenomenal, and is hugely appreciated!

**Comment by mohesles** [comment]
I have an Amcrest NVR so I can test if needed. As mentioned on python-amcrest. I have tested and validated that the following URL formats work for different channels (cameras):

http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=0
http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=1
http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=2
http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=3

In addition to that there is a substream parameter that I also tested:

http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=0&subtype=1
http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=1&subtype=1
http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=2&subtype=1
http://user:pass@NVRs_IP/cgi-bin/snapshot.cgi?channel=3&subtype=1

**Comment by pnbruckner** [comment]
@mohesles thanks!

But as I mentioned above, it's more than just the snapshot & RTSP URL. Currently there are 24 different AmcrestCamera methods in the amcrest package used by the HA amcrest component (including what I use in my custom version and what I'm ultimately planning to submit to the standard component.) Besides enhancing them to deal with a channel number (which is not as straightforward as you might expect/hope), for some of them it might not even make sense (e.g., does the NVR "pass through" those commands, and/or do they apply more to the NVR than the cameras themselves???)

I'm not saying I won't ever attempt what is being discussed, and I appreciate the data and the offer to test, but implementing something will probably require some experimenting (since the API doc leaves much to be desired), and that's really hard to do without direct access to an NVR (which I don't have, and probably don't plan to get.)

**Comment by GaryOkie** [comment]
Hi @pnbruckner - 

The API for the NVR's is the same as for the cameras.   Cameras don't need a channel# to be specified, but it's included for compatibility if an NVR is being accessed instead of a camera directly.   While I can't say I fully understand exactly how an NVR manages the traffic, it sure does appear to simply pass-through the calls to each camera.   

The only situation I can think of that is unique to an NVR is the use of channel 0.     If the NVR is configured to enable this unique channel, it will provide a matrix display of all enabled channels returned in 1 snapshot.    I'm pretty sure channel 0 will not work with RTSP, and even the docs say channel starts at 1.

I have not configured my NVR for this, and if I select channel 0 in a snapshot, it grabs channel 1 instead.

I had indeed  hoped it would be a straightforward update to change a hardcoded channel=1 in the calls to a variable.   Yeah, all 24 of them! :)

**Comment by GaryOkie** [comment]
FWIW - I originally was using the generic platform.   I was connecting to the NVR directly specifying the channel.    It made absolutely no difference changing the url's to connect to the camera's directly.   Perhaps further evidence of their API equivalency.
```
    stream:

    camera 1:   
  - platform: generic
    name: "NVR camera 1"
    stream_source:   "rtsp://uid:pw@192.168.2.20:554/cam/realmonitor?channel=1&subtype=0"
    still_image_url: "http://uid:pw@192.168.2.20/cgi-bin/snapshot.cgi?channel=1"

    camera 2:   
  - platform: generic
    name: "NVR camera 2"
    stream_source:   "rtsp://uid:pw@192.168.2.20:554/cam/realmonitor?channel=2&subtype=0"
    still_image_url: "http://uid:pw@192.168.2.20/cgi-bin/snapshot.cgi?channel=2"

    camera 3:   
  - platform: generic
    name: "NVR camera 3"
    stream_source:   "rtsp://uid:pw@192.168.2.20:554/cam/realmonitor?channel=3&subtype=0"
    still_image_url: "http://uid:pw@192.168.2.20/cgi-bin/snapshot.cgi?channel=3"

    camera 4:   
  - platform: generic
    name: "NVR camera 4"
    stream_source:   "rtsp://uid:pw@192.168.2.20:554/cam/realmonitor?channel=4&subtype=0"
    still_image_url: "http://uid:pw@192.168.2.20/cgi-bin/snapshot.cgi?channel=4"
```
Note:   newer firmware requires authentication to be digest, not basic.  So this was also added:
    verify_ssl: false
    authentication: digest

And lastly, I am perfectly fine connecting to the cameras directly using amcrest since my network now allows it.   Having a new channel option provides some modest additional network flexibility and security (single pw), so really no big deal.   

Just thought this option could be added if it were easy enough.    Realistically, as long as all 24 calls default to channel 1 instead of being hardcoded, I would think that the impact would be zero to existing non-NVR users, and anyone using an NVR could just try it and report findings.

**Comment by pnbruckner** [comment]
I think you might be missing the point. There are many commands used by the HA amcrest component that _don't_ have or allow a channel option. That's where it gets tricky.

**Comment by GaryOkie** [comment]
youre right.    I missed this point.   I have no idea how a command without a channel needs to be passed on to a particular camera.

I'll close this request.    many other more important things to do.

**Comment by pnbruckner** [comment]
If you don't mind I'll reopen. These are good ideas, and I'd like to eventually get something working, but as I said, I don't have the devices to experiment with, and it will take some experimenting because of the issues I stated, and the fact the doc is a bit vague. Also the way the amcrest package is currently written it will be a challenge without significantly changing its API, which maybe should be done anyway. But I'm focusing on other things for now. Leaving this open will remind me to eventually come back to this. Maybe at that time you or others will be willing to do some of the experimenting for me. 😄

**Comment by GaryOkie** [comment]
Sure - I don't mind at all reopening it!   And of course, I'll be more than happy to help test.

**Comment by akram** [comment]
Hi @pnbruckner ,

I would be interested in working on this one.

**Comment by pnbruckner** [comment]
@akram thanks. But I'm no longer supporting this custom component/integration, mainly because most of the enhancements I worked on I've submitted to the standard component. And right now I'm actually focusing on resolving some nagging issues with that component and the amcrest package it uses.

**Comment by akram** [comment]
@pnbruckner that's ok, but I want to contribute even if you can't provide the support anymore. I saw that this would also require improvements on python-amcrest.

**Comment by pnbruckner** [comment]
@akram I suggest you start a new custom integration. You can, of course, use whatever of what I did as you like. I doubt, however, that the custom integration in this repo would even work anymore.

I also wonder how easily the current amcrest integration could be expanded to support NVRs. It might make more sense to have a separate amcrest_nvr integration.

**Comment by akram** [comment]
@pnbruckner can you give me more details about what you call:
- the custom integration
- the current amcrest integration
- amcrest package 

I was interrested in contributing in https://github.com/home-assistant/home-assistant/tree/dev/homeassistant/components/amcrest 

I am just using this issue as a previous discussion started here and that the exposed ideas seems correct to me.

regarding amcrest_nvr , it does not make sense to have a specific integration for this one, as the only difference with amcrest existing integration would be the ability to pass/propagate the "channel" . In the existing integration it is not present in the constructors, but it is set in other places in the code and has default value of 1 if passed as None.

**Comment by pnbruckner** [comment]
@akram thanks for the clarification. It wasn't clear what your intention was. I assumed you meant making enhancements to the implementation in this repo.

Those are some fairly basic HA development terms...

An integration provides functionality related to some type of device, and creates entities of one or more types of components by providing corresponding platform code.

A custom integration is an integration that is not part of the main HA distribution and is intended to reside in a user's `<config>/custom_components` directory.

The "current amcrest integration", aka the "standard amcrest integration", is the one which _is_ part of the main HA distribution (located at the link you gave above.)

The amcrest package is the package distributed via [pypi.org](https://pypi.org/project/amcrest/) that the amcrest integration uses. The source can be found [here](https://github.com/tchellomello/python-amcrest).

I'm not sure I agree with your conclusion. There are at least some camera commands that the integration uses that do not accept a channel parameter. At the very least it won't be a "slam dunk." But I don't want to discourage you. I know this feature has been wanted, so good luck!

**Comment by akram** [comment]
@pnbruckner it is not discouraging. I already had a look at the code and it is doable for sure. I know https://github.com/tchellomello/python-amcrest code base and I already opened an issue there for a specific feature that can benefit also HA, but not required yet.

What would be the next step? I think I should start opening an iusse with https://github.com/home-assistant/home-assistant/ targeting the amcrest component and I will reference this issue just for discussion.

Then I will push a PR asap. I have an NVR so I can test the integration. My only problem is the test environment. Cause, I can't override my current remote installation, so I will make another install and create SSH tunnel to have access to the camera.
